### PR TITLE
Add Qt integration to sentry

### PR DESCRIPTION
**Pull request by MartinDelille**
Specify library name and version:  **sentry/0.4.9**

See the doc here: https://docs.sentry.io/platforms/native/guides/qt/
---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/sentry-native/all/conanfile.py [hunk 1]
```diff
@@ -73,6 +75,7 @@ def _configure_cmake(self):
         self._cmake.definitions["SENTRY_ENABLE_INSTALL"] = True
         self._cmake.definitions["SENTRY_TRANSPORT"] = self.options.transport
         self._cmake.definitions["SENTRY_PIC"] = self.options.get_safe("fPIC", True)
+        self._cmake.definitions["SENTRY_INTEGRATION_QT"] = self.options.get_safe("qt", False)
```
Comment (ericLemanissier) [inline recipes/sentry-native/all/conanfile.py:78]:
```suggestion
        self._cmake.definitions["SENTRY_INTEGRATION_QT"] = self.options.qt
```
no need for get_safe if the option is never deleted

**Review by ericLemanissier — COMMENTED** [review]
You could add
```
if self.options.qt:
    self.requires("qt/....")
```

**Comment by MartinDelille** [comment]
@ericLemanissier I personnaly don't use the Qt recipe from CCI yet (I'd love to do so but it isn't very clear for me how to do it, especially with my local environment and Qt Creator). Won't this cause trouble for my setup ?

**Comment by ericLemanissier** [comment]
If you require another qt recipe in your recipe (`qt/...@bincrafters/stable` for example), then this qt package will be used instead of CCI's package https://docs.conan.io/en/latest/using_packages/conanfile_txt.html#overriding-requirements

If you are using a qt coming from somewhere else than conan (manually built, official qt binary, or provided by some other package manager), it's true that it could be a problem.

**Comment by MartinDelille** [comment]
This is quite out of topic but I'll be happy to know how people use the CCI qt recipe with Qt Creator for example. Do I need to run conan and then add the kit manually ?

**Comment by madebr** [comment]
> You could add
> 
> ```
> if self.options.qt:
>     self.requires("qt/....")
> ```

The recipe is missing `self.requires("qt/...")`.
Does this recipe actually build when no qt is installed and configured with `-o sentry-native:qt=True`?

**Review by SSE4 — COMMENTED** [review]
please clarify if Qt integration requires Qt dependency to build/run. I am not familiar with sentry, so it's hard to judge for me right away.
(e.g. I can imagine it may not require an actual Qt if it only generates some headers containing `#include <QtWidgets>` or similar code)

**Comment by MartinDelille** [comment]
The recipe does indeed need Qt to build if the option is set https://github.com/getsentry/sentry-native/blob/master/CMakeLists.txt#L462-L475

Unfortunately when trying to add the requirement, `cmake` is unable to find Qt when building sentry-native...

I pushed the update recipe with qt option set to true to share my issue.

**Comment by mjvankampen** [comment]
In my opinion the CCI Qt recipe is not a full drop in replacement for the official Qt binaries (mostly qml related). So I wouldn't impose Qt as a requirement.

File: recipes/sentry-native/all/conanfile.py [hunk 2]
```diff
@@ -59,11 +61,18 @@ def requirements(self):
             raise ConanInvalidConfiguration("crashpad not available yet in CCI")
         elif self.options.backend == "breakpad":
             raise ConanInvalidConfiguration("breakpad not available yet in CCI")
+        if self.options.qt:
+            self.requires("qt/5.15.2")
+            self.requires("openssl/1.1.1j")
```
Comment (prince-chrismc) [inline recipes/sentry-native/all/conanfile.py:66]:
```suggestion
            self.requires("openssl/1.1.1k")
```

File: recipes/sentry-native/all/conanfile.py [hunk 3]
```diff
@@ -59,11 +61,18 @@ def requirements(self):
...
+            if tools.Version(self.version) < "0.4.5":
+                raise ConanInvalidConfiguration("Qt integration available from version 0.4.5")
 
     def source(self):
         tools.get(**self.conan_data["sources"][self.version])
         extracted_dir = self.name + "-" + self.version
         os.rename(extracted_dir, self._source_subfolder)
+        for patch in self.conan_data.get("patches", {}).get(self.version, []):
```
Comment (prince-chrismc) [inline recipes/sentry-native/all/conanfile.py:74]: Needs to be in build

**Comment by prince-chrismc** [comment]
@mjvankampen Its not 100% drop in just yet but perhaps you can add logs to the existing issues and detail the short comings.
If you know of any possible solutions this way others will be able to build and hopefully solve the issues.

**Comment by MartinDelille** [comment]
> @mjvankampen Its not 100% drop in just yet but perhaps you can add logs to the existing issues and detail the short comings.
> If you know of any possible solutions this way others will be able to build and hopefully solve the issues.

I spend the last day to try to migrate to qt/5.15.2 from conan and I have a blocking issue: #5487

**Review by prince-chrismc — APPROVED** [review]

**Review by prince-chrismc — APPROVED** [review]

**Review by SSE4 — DISMISSED** [review]

**Review by mathbunnyru — CHANGES_REQUESTED** [review]
Patches for 0.4.7 and 0.4.8 are the same, right?
Let's use one file instead of two.

**Comment by prince-chrismc** [comment]
@mathbunnyru The patches are on different line so two files is required ='(

@MartinDelille You PR has merge conflicts, please update your branch

**Review by uilianries — DISMISSED** [review]

**Review by mathbunnyru — DISMISSED** [review]

**Comment by SSE4** [comment]
from that I see, it reports merge conflicts which don't exist (at least GitHub doesn't report them):
```
+ git merge --no-ff 2f71c2f10dfbe4a87595bdc619aeef8262a9e6d0
Removing recipes/vorbis/all/patches/link-libm.patch
Removing recipes/srt/all/patches/002-cmake-link-targets.patch
Removing recipes/srt/all/patches/001-cmake-link-targets.patch
Auto-merging recipes/sentry-native/all/conanfile.py
Auto-merging recipes/sentry-native/all/conandata.yml
CONFLICT (content): Merge conflict in recipes/sentry-native/all/conandata.yml
Removing recipes/pkgconf/all/patches/0002-c89-compatible.patch
Automatic merge failed; fix conflicts and then commit the result.
```
probably, bug in CI? @jgsogo @danimtb

**Comment by jgsogo** [comment]
@MartinDelille , please, push an empty commit to your branch:

```
git commit -m "touch" --allow-empty
git push
```

**Review by mathbunnyru — APPROVED** [review]

File: recipes/sentry-native/all/conanfile.py [hunk 4]
```diff
@@ -16,20 +16,22 @@ class SentryNativeConan(ConanFile):
...
         "transport": ["none", "curl", "winhttp"],
+        "qt": [True, False],
     }
     default_options = {
         "shared": False,
         "fPIC": True,
         "backend": "inproc",
-        "transport": "curl"
+        "transport": "curl",
+        "qt": False,
```
Comment (madebr) [inline recipes/sentry-native/all/conanfile.py:34]: Is qt integration by default useful?

**Review by madebr — APPROVED** [review]

File: recipes/sentry-native/all/conanfile.py (hunk 4 above)
Comment (prince-chrismc) [reply recipes/sentry-native/all/conanfile.py:34]: Yes but at the expense of current users who do no require it

https://docs.sentry.io/platforms/native/guides/qt/#install

Looks off by default from upstream

**Review by prince-chrismc — APPROVED** [review]

File: recipes/sentry-native/all/conanfile.py (hunk 4 above)
Comment (madebr) [reply recipes/sentry-native/all/conanfile.py:34]: buildroot has some kind of ci system where it builds random combinations to find regressions.
Something similar where non-default option combinations are built would be useful.

**Pull request merged**